xnmt/inference.py

```diff
@@ -116,28 +134,33 @@ def __call__(self, generator, src_file=None, trg_file=None, candidate_id_file=No
       ref_scores = [-x for x in ref_scores]
 
     # Perform generation of output
-    with io.open(args["trg_file"], 'wt', encoding='utf-8') as fp:  # Saving the translated output to a trg file
-      src_ret=[]
-      for i, src in enumerate(src_corpus):
-        # This is necessary when the batcher does some sort of pre-processing, e.g.
-        # when the batcher pads to a particular number of dimensions
-        if self.batcher:
-          self.batcher.add_single_batch(src_curr=[src], trg_curr=None, src_ret=src_ret, trg_ret=None)
-          src = src_ret.pop()[0]
+    if args["mode"] != 'score':
+      with io.open(args["trg_file"], 'wt', encoding='utf-8') as fp:  # Saving the translated output to a trg file
+        src_ret=[]
+        for i, src in enumerate(src_corpus):
```

Comment:
How about we just step over each line in the target file here, we then read in the first column specifying the index to decide which source sentence we want to encode, then use it to score the target sentence. To read in the target sentence, we can use `generator.trg_reader.read_sent(val)`, which reads in a single sentence. Right now in `input.py` there is no function to read a single sentence, but I think this is something that is obviously necessary, so I think it'd be nice to make all of the `read_sents` calls basically call out to `read_sent()`. I think this also might help reduce code duplications across the various subclasses of `InputReader`.

Reply:
input_reader calls a iterate_filtered(filename, filter_id) function on the files. I am not sure what is the best way to separate a read_sent function from this. Do you think we should make read_sents(filename) function call iterate_filtered(filename), and make iterate_filtered(filename) call read_sent() for each line in the file?

Reply:
Yeah, that sounds right. `read_sent` will take in a plain text string, and convert it into the appropriate `xxxInput` data structure.